projects/dqn/src/rl_experiment.py

```diff
@@ -0,0 +1,278 @@
+            self.dataset.observation_space.shape
+
+        # Num steps required for algorithm schedulers
+        self.num_steps = config.get("num_steps", None)
+        if self.num_steps is None:
+            raise ValueError("Must define num_steps")
+        algorithm_args["num_steps"] = self.num_steps
+
+        # Define reward clipping, if existing
+        self.reward_clip = config.get("reward_clip", None)
+
+        # Initialize algorithm
+        algorithm_class = config.get("algorithm_class", None)
+        if algorithm_class is None:
+            raise ValueError("Must define algorithm_class")
+
+        self.algorithm = algorithm_class(
+            **algorithm_args,
+            device=self.device
+        )
+
+        self.replay_buffer = ReplayBuffer()
+        self.buffer_batch_size = config.get("buffer_batch_size", 64)
+        self.gradient_steps_per_epoch = config.get("gradient_steps_per_epoch", 16)
+
+        if self.dataset is None:
+            raise ValueError("Dataset must be initialized beforehand.")
+
+        self.data_collection_loader = self.create_data_collection_loader()
+
+        self.num_episodes_per_epoch = config.get("num_episodes_per_epoch", 1)
+
+    def collect_data(self, num_episodes):
+        """Return the mean of episode rewards"""
+
+        episode_rewards = Counter()
+        # how to run one step in data collection?
+        for episode in range(num_episodes):
+            for experience in self.data_collection_loader:
+                observation, action, reward, done, next_observation, info = experience
+                # Clip rewards and other transformations from the agent side
+                if self.reward_clip is not None:
+                    min_reward, max_reward = self.reward_clip
+                    reward = max(min_reward, min(reward, max_reward))
+                episode_rewards[episode] += reward
+                # Append to buffer
+                self.replay_buffer.append(experience)
+
+        return np.mean(list(episode_rewards.values()))
+
+    def pre_epoch(self):
+        pass
+
+    def post_epoch(self):
+        pass
+
+    def run_epoch(self):
+
+        self.pre_epoch()
+
+        # Collect Experiences
+        ret = {}
+        t0 = time.time()
+        ret["average_reward"] = self.collect_data(self.num_episodes_per_epoch)
+
+        # Train
+        t1 = time.time()
+        ret["average_loss"] = self.train_epoch()
+        t2 = time.time()
+
+        self.post_epoch()
+
+        self.logger.debug("data collection time: %s", t1 - t0)
+        self.logger.debug("training time: %s", t2 - t1)
+        self.logger.debug("---------- End of run epoch ------------")
+        self.logger.debug("")
+
+        self.current_epoch += 1
+        return ret
+
+    def train_epoch(self):
+        # Train loader must be recreated at each step due to different sample subset
+        train_loader = self.create_train_loader()
+        epoch_loss = 0
```

Comment:
If you use a GPU tensor for `epoch_loss` then you can keep the loss returned by `gradient_step` also on GPU and avoid the GPU->CPU transfer in the loop
```
 epoch_loss = torch.tensor(0, device=self.device)
```

Reply:
Good idea, but only moving an integer shouldn't be that costly. In this case would rather have my gradient step in the algorithm to return an int as the loss and not a tensor so it doesn't get messy when logging or debugging.